File: tapo-py/examples/tapo_p300.py
"""P300 and P306 Example"""

import asyncio
import os

from tapo import ApiClient


async def main():
    tapo_username = os.getenv("TAPO_USERNAME")
    tapo_password = os.getenv("TAPO_PASSWORD")
    ip_address = os.getenv("IP_ADDRESS")

    client = ApiClient(tapo_username, tapo_password)
    power_strip = await client.p300(ip_address)

    device_info = await power_strip.get_device_info()
    print(f"Device info: {device_info.to_dict()}")

    print("Getting child devices...")
    child_device_list = await power_strip.get_child_device_list()
    print(f"Found {len(child_device_list)} plugs")

    for index, child in enumerate(child_device_list):
        print(f"=== ({index + 1}) {child.nickname} ===")
        print(f"Device ID: {child.device_id}")
        print(f"State: {child.device_on}")

        plug = await power_strip.plug(device_id=child.device_id)

        print("Turning device on...")
        await plug.on()

        print("Waiting 2 seconds...")
        await asyncio.sleep(2)

        print("Turning device off...")
        await plug.off()

        print("Waiting 2 seconds...")
        await asyncio.sleep(2)


if __name__ == "__main__":
    asyncio.run(main())

File: tapo/examples/tapo_p300.rs
/// P300 and P306 Example
use std::{env, thread, time::Duration};

use log::info;
use tapo::{ApiClient, Plug};

mod common;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    common::setup_logger();

    let tapo_username = env::var("TAPO_USERNAME")?;
    let tapo_password = env::var("TAPO_PASSWORD")?;
    let ip_address = env::var("IP_ADDRESS")?;

    let power_strip = ApiClient::new(tapo_username, tapo_password)
        .p300(ip_address)
        .await?;

    let device_info = power_strip.get_device_info().await?;
    info!("Device info: {device_info:?}");

    info!("Getting child devices...");
    let child_device_list = power_strip.get_child_device_list().await?;
    info!("Found {} plugs", child_device_list.len());

    for (index, child) in child_device_list.into_iter().enumerate() {
        info!("=== ({}) {} ===", index + 1, child.nickname);
        info!("Device ID: {}", child.device_id);
        info!("State: {}", child.device_on);

        let plug = power_strip.plug(Plug::ByDeviceId(child.device_id)).await?;

        info!("Turning device on...");
        plug.on().await?;

        info!("Waiting 2 seconds...");
        thread::sleep(Duration::from_secs(2));

        info!("Turning device off...");
        plug.off().await?;

        info!("Waiting 2 seconds...");
        thread::sleep(Duration::from_secs(2));
    }

    Ok(())
}
